frontend/src/libs/importExportUtils.js

```diff
@@ -328,3 +329,306 @@ export function dataToFile(formatters, dataFormat, filePrefix = "") {
+        // We want to flatten a lot of the data in `assignments` and only include the information
+        // we need.
+        const assignmentsForSpreadsheet = assignments.map((assignment) => ({
+            first_name: assignment.applicant.first_name,
+            last_name: assignment.applicant.last_name,
+            utorid: assignment.applicant.utorid,
+            email: assignment.applicant.email,
+            position_code: assignment.position.position_code,
+            start_date: assignment.start_date,
+            end_date: assignment.end_date,
+            contract_template: assignment.contract_override_pdf
+                ? null
+                : assignment.position.contract_template.template_name,
+            contract_override_pdf: assignment.contract_override_pdf,
+            hours: assignment.hours,
+            active_offer_status: assignment.active_offer_status,
+            active_offer_recent_activity_date:
+                assignment.active_offer_recent_activity_date,
+            wage_chunks: assignment.wage_chunks.map((chunk) => ({
+                hours: chunk.hours,
+                rate: chunk.rate,
+                start_date: chunk.start_date,
+                end_date: chunk.end_date,
+            })),
+        }));
+        return [
+            [
+                "Last Name",
+                "First Name",
+                "UTORid",
+                "Email",
+                "Position Code",
+                "Start Date",
+                "End Date",
+                "Hours",
+                "Contract Template",
+                "Contract Override PDF",
+                "Offer Status",
+                "Recent Activity Date",
+                "",
+                "Number of Pay Periods",
+                ...createPayPeriodHeaders(assignmentsForSpreadsheet),
+            ],
+        ].concat(
+            assignmentsForSpreadsheet.map((assignment) => [
+                assignment.last_name,
+                assignment.first_name,
+                assignment.utorid,
+                assignment.email,
+                assignment.position_code,
+                formatDateForSpreadsheet(assignment.start_date),
+                formatDateForSpreadsheet(assignment.end_date),
+                assignment.hours,
+                assignment.contract_template,
+                assignment.contract_override_pdf,
+                assignment.active_offer_status,
+                undefined,
+                assignment.active_offer_recent_activity_date,
+                ...formatWageChunksToList(assignment.wage_chunks),
+            ])
+        );
+    },
+};
+
+/**
+ * A factory function which produces corresponding prepareData function,
+ *  for each type of object (possibly needs filtering or associated session).
+ *
+ * @param {string} objectType
+ * @param {Function | null} objectFilter
+ * @param {Session | null} session
+ * @returns {Function}
+ */
+export function prepareDataFactory(
```

Comment:
This is too general. We don't want a general factory function. Only one for the specific case where we need it.

Reply:
There are two objects having special cases among five objects. 

1. `Ddah` requires filtering with `selectedDdahIds`
2. `Assignment` requires filtering with `selectedAssignmentIds` and also has different `prepareMinimal` parameters than all the other objects

Just want to confirm, do we want only one factory function for there two objects? Or do we need one factory function for each of them

Reply:
Nope. A separate factory function for each of those two.

Reply:
Also, I noticed that in the original code, only` Ddah` made `prepareDdahsSpreadsheet` and passed it to `dataToFile`, the other objects do the parsing to spreadsheet inside `dataToFile` parameter. In order for consistency, I extracted all object's parsing to spreadsheet out and created a constant `prepareSpreadsheet`, which functions similarly as `prepareMinimal`. I wonder do we want to keep this change or completely revert to the original code on Master?

Reply:
It's a nice idea to break them all out. It would be nice if we converted the file to TypeScript so that we could have some safety guarantees.

Why are all the helper functions, like `flattenDuties` exported? I would hope they are only used in prepare spreadsheet function...